### ISSUE TITLE
[dart2wasm] JS errors can't be caught using an interop type

### ISSUE THREAD

**Issue by srujzs** [dart-lang/sdk#55481]
```dart
import 'dart:js_interop';

@JS()
external void eval(String code);

@JS()
external void throwFunction();

void main() {
  eval('''
    self.throwFunction = function() {
      throw new Error('hello world');
    }
  ''');
  try {
    throwFunction();
  } on JSObject catch (e) {
    print(e);
  }
}
```

DDC and dart2js catch the JS error as it is a subtype of `interceptors.JavaScriptObject` (and therefore a subtype of `JSObject`). dart2wasm does not, as its implementation of JS errors is not a subtype of `JSObject`: https://github.com/dart-lang/sdk/blob/ad95bb6f89eb3aa36b129320fba47c29574b7084/sdk/lib/_internal/wasm/lib/errors_patch.dart#L40. A proposal is to make that implementation a subtype of `JSValue`, but it's not clear if a Wasm instruction exists to convert the error to an `externref` that can then be boxed. Making this a subtype of `JSValue` also has the other issue that if we replaced `JSObject` with `JSString`, dart2wasm would catch the error (because `JSString` is a `JSValue`) whereas DDC and dart2js will not.

For now, users should avoid catching JS errors using an interop type for consistency.

**Comment by osa1** [comment]
> but it's not clear if a Wasm instruction exists to convert the error to an externref that can then be boxed

This will be possible with https://github.com/WebAssembly/exception-handling/pull/269. Chrome already supports it, but the latest Firefox (125.0.1) doesn't.

**Comment by mkustermann** [comment]
If this becomes a blocker for some customers as production error reporting may not contain any meaningful information, then we probably have to find some solution for this.

If we're not willing to generally raise the minimum version of browsers supported (@kevmoo ), we could offer it as a compiler option to generate new vs old instructions so some customers can raise their minimum version and get better errors.

If important customers still want to support older browsers, one option may be to push the responsibility of capturing the error to the JS side: So wrap every JS code we call in a `try { return <original code> } catch (e) {savedError = e; throw e;}`. If we then catch a JS error in wasm, we can call out to JS to access the `savedError`. The question how expensive it would be to do this.

**Comment by srujzs** [comment]
I do think this will become relevant for internal apps.

>  So wrap every JS code we call in a try { return <original code> } catch (e) {savedError = e; throw e;}.

Is there a better way to do this than on every JS member that we emit? A redirecting method would probably be too slow, but having it on every JS member would be a lot of extra code in the JS module.

**Comment by osa1** [comment]
> A proposal is to make that implementation a subtype of JSValue, but it's not clear if a Wasm instruction exists to convert the error to an externref that can then be boxed

Yes, we're doing this in https://dart-review.googlesource.com/c/sdk/+/469062 which is ready for merging.

With the CL, we catch the JS exceptions as `externref` and box them as `_JavaScriptError`. This improves error reporting as we use JS `toString` of the error and JS stack traces (instead of Dart stack traces at the catch site).

> Making this a subtype of JSValue also has the other issue that if we replaced JSObject with JSString, dart2wasm would catch the error (because JSString is a JSValue) whereas DDC and dart2js will not.

I don't have an opinion on this but this is the main question is this issue, so assigning @srujzs.

For reference, this is the current test we use to decide whether to catch JS exceptions or not: https://github.com/dart-lang/sdk/blob/ced923c8ca0a6be214407bbc744473b5a1ff267c/pkg/dart2wasm/lib/code_generator.dart#L5168-L5180

**Comment by srujzs** [comment]
Thanks Omer!

Having a better `toString` and stack trace is great (and to be clear, this CL is a strict improvement), but yeah, this is tricky. There are a few concerns I have with this:

- There's no way to check that the exception is a JS error beyond looking at the string/trace as `_JavaScriptError` is not exposed. With the JS compilers, this naturally is not an issue as the value itself is the JS error.
- Even if we were to expose the error, we'd need to have a getter on it that returns a `JSAny` which wraps the underlying `externref` so users can use that value (probably for type tests).
- If we exposed the error with a getter, users would be forced to write platform-specific code like (assume `isJSAny` exists here):
```dart
...
} catch (e) {
  JSAny? jsErrorValue;
  if (isWasm) {
    if (e is JavaScriptError) {
      jsErrorValue = e.jsError;
    }
  } else if (e.isJSAny) {
    jsErrorValue = e as JSAny;
  }
  if (jsErrorValue != null) {
    // JS error handling
  } else {
    // common Dart exception handling
  }
}
```

If we instead used `JSValue`, we still come across the above issue I mentioned, but we have better error checking:

```dart
} catch (e) {
  if (e.isJSAny) {
    // JS error handling
  } else {
    // common Dart exception handling
  }
}
```

It also aligns with the general discrepancy in interop types we've communicated to users before. We can even expand `invalid_runtime_check_with_js_interop_types` to catch instances like `} on <interop type> catch (e) {` and recommend users do a combination of `isJSAny` and `isA` checks instead.

**Comment by mkustermann** [comment]
> A proposal is to make that (_JavaScriptError) implementation a subtype of JSValue

Just to clarify:
```dart
extension type Foo(JSObject _) implements JSObject {}
extension type Bar(JSObject _) implements JSObject {}

@JS()
void xxx();

main() {
  try {
     xxx();
  } on Foo catch (Foo e) {
    print(e);
  } on Bar catch (Bar e) {
    print(e);
  }
}
```

It would be problematic to allow this because JS interop types like `Foo` are extension types and are therefore only syntactic sugar with the runtime values be the representation type (which is compiler-defined, in wasm the `JSValue`). Both semantics for this are problematic:

* If we say it's only syntactic sugar then `on Foo` becomes `on JSValue` and we catch all JS errors in the first catch and wrap/downcast to `Foo` - even for errors that are not `Foo`s - clearly not user intention
* If we say it's catching it as JS error and internally performs a `isA<Foo>` check, then this would be in stark contrast to the way we perform `if (obj is Foo)` test, which are basically just `if (obj is JSValue)` checks.

=> So based on that it seems it would be better to issue lints/errors on such catch clauses.

In an ideal world it would seem cleanest to represent non-Dart errors as a foreign `JavaScriptError` that one can catch. Exposing a getter for the underlying foreign error object that was thrown & one can use that to perform foreign type tests on. - just like we'd represent any foreign JS object via a Dart wrapper object.

Looking at what dart2js does [in this code](https://github.com/dart-lang/sdk/blob/2b7ca1ce7b67df056904a02eac837c24d008b018/sdk/lib/_internal/js_runtime/lib/js_helper.dart#L1388) it seems that

* when we throw in Dart, we make an `Error` with a `dartException` property 
* when we catch, we can identify those Dart exceptions via that property
* if it's not a Dart error, we check if it's JS stackoverflow error, etc and convert to equivalent Dart errors
* otherwise, the `error` is just the JS object and accesses to it are probably using interceptors

So theoretically dart2js knows if the error originates from Dart or not, and could therefore construct `JavaScriptErrors` that wrap the foreign thrown errors. Though I guess this would be a breaking change (just like it would be breaking to wrap foreign JS objects in a wrapper Dart object)?

So based on that maybe it's best to follow this pattern from dart2js
* check for certain JS errors (e.g. null thrown, stack overflow, etc) and convert them to corresponding Dart errors.
* otherwise use the JS object as error value (which in dart2wasm case would be `JSValue`s).

And correct interop code should use `} on JSAny catch (JSAny e) { <perform type checks on `e`> }` - then it will work correctly in both dart2js and dart2wasm.

I guess this is an elaborate version of what @srujzs said above?

**Comment by srujzs** [comment]
> So theoretically dart2js knows if the error originates from Dart or not, and could therefore construct JavaScriptErrors that wrap the foreign thrown errors. Though I guess this would be a breaking change (just like it would be breaking to wrap foreign JS objects in a wrapper Dart object)?

Yeah, I agree that an ideal model would have the JS error properly boxed with a single class across all compilers. It makes it obvious that the result is boxed and the user needs to do appropriate checks outside of the `catch`. But it is potentially a large breaking change.

> And correct interop code should use } on JSAny catch (JSAny e) { <perform type checks on e> } - then it will work correctly in both dart2js and dart2wasm.

Kind of, but the user can't type the error without determining if it actually is a `JSAny` using e.g. `isJSAny`. In the JS compilers, that would catch *all* errors, Dart or otherwise, as `JSAny` is just `Object`.

> I guess this is an elaborate version of what @srujzs said above?

Yes, but you added some important details. :)

**Comment by mkustermann** [comment]
> Kind of, but the user can't type the error without determining if it actually is a JSAny using e.g. isJSAny. In the JS compilers, that would catch all errors, Dart or otherwise, as JSAny is just Object.

Ah yes, that's true.

Is the `isJSAny` functionality we need to add?

**Comment by srujzs** [comment]
> Is the isJSAny functionality we need to add?

Yup, that and `isJSObject` would be useful. We have a tracking bug that's still on me: https://github.com/dart-lang/sdk/issues/56905.

**Comment by srujzs** [comment]
A quick note from a discussion between me and Martin: with `isJSAny`, the treatment of thrown primitives is a bit odd. In the JS compilers, Dart primitives *are* JS primitives, so `isJSAny` will always return true for a Dart `String`. On dart2wasm, this isn't true and there is a runtime type difference between a JS primitive and a Dart primitive.

Therefore, primitives must be treated in both the `if` and `else` blocks of `isJSAny`, with the caveat that the primitive treatment probably shouldn't differ between the two blocks (so that it can be consistent across all platforms). This isn't new/surprising for `isJSAny`, but in the context of this bug, this does mean the user can't tell if the primitive was thrown from JS in the JS compilers, whereas you can with dart2wasm.

This should be documented on `isJSAny` and we should try and make that behavior obvious to users (maybe through our existing lint?).

**Comment by kevmoo** [comment]
@srujzs – agree if we add `isJSAny` we should warn about it.

Sadly, "linting" on a known static type `String`, etc isn't super helpful.

**Comment by osa1** [comment]
We discussed this in a few meetings. To summarize the plans here:

We'll remove `_JavaScriptError` and only catch JS errors as ~~`Object?`. (e.g. `on Object? catch ...`, or `on dynamic catch ...`)~~ correction: we'll catch JS errors as `Object`, `Object?`, or `dynamic`. `null` and `undefined` values will be converted to a non-interop type. dart2js uses a private `NullThrownFromJavaScriptException` class for this, we can do the same.

We'll check interop types in `on ...` in a lint check, as they won't be catching anything in dart2wasm.

JS objects caught in Wasm will be represented as `JSValue`. So that means they can be used as any JS interop type like `JSString`, `JSObject` etc.

Currently `_JavaScriptError` implements `Error` and has a field for the stack trace. Users can use this by casting the caught object to `Error` or using it as `dynamic`. This won't work anymore as we can't attach a stack trace to `JSValue`s.

~~I'll implement the changes in dart2wasm.~~ Being implemented in: https://dart-review.googlesource.com/c/sdk/+/479640
